internals/overlord/state/identities.go

```diff
@@ -266,17 +324,43 @@ func (s *State) Identities() map[string]*Identity {
 
 // IdentityFromInputs returns an identity with the given inputs, or nil
 // if there is none.
-func (s *State) IdentityFromInputs(userID *uint32) *Identity {
+//
+// Identity priority:
+//  1. If both username and password are provided, the function attempts to
+//     match a basic type identity. The userID is ignored in this case. If
+//     a matching username is found but the password verification fails, nil
+//     is returned immediately.
+//  2. If username and password are not both provided, the function attempts to
+//     match a local type identity using the userID.
+//
+// If no matching identity is found for the given inputs, nil is returned.
+func (s *State) IdentityFromInputs(userID *uint32, username, password string) *Identity {
 	s.reading()
 
-	for _, identity := range s.identities {
-		switch {
-		case identity.Local != nil && userID != nil:
-			if identity.Local.UserID == *userID {
+	switch {
+	case username != "" || password != "":
+		// Prioritize username/password if provided, because they come from HTTP
+		// Authorization header, a per-request, client controlled property. If set
+		// by the client, it's intentional, so it should have a higher priority.
+		for _, identity := range s.identities {
+			if identity.Basic != nil && identity.Name == username {
+				crypt := sha512_crypt.New()
+				err := crypt.Verify(identity.Basic.Password, []byte(password))
```

Comment:
How about we put `passwordBytes := []byte(password)` above the for loop to avoid allocating/copying each time (maybe Go's compiler is smart enough to avoid this, but I doubt it).

Reply:
Yes, it doesn't look good. I will avoid allocation in each loop.

Out of curiosity, I did a benchmark test and it seems it doesn't make a difference, they are more or less the same and even allocation might be faster sometimes. But I agree, it doesn't read very nice because it at least feels expensive :) I will move it out of the loop.